examples/bridge/sn/src/dojo_bridge.cairo

```diff
@@ -0,0 +1,192 @@
+        #[key]
+        sender: ContractAddress,
+        recipient: felt252,
+        amount: u256
+    }
+
+    mod Errors {
+        const CALLER_IS_NOT_OWNER: felt252 = 'Bridge: caller is not owner';
+        const INVALID_L1_ORIGIN: felt252 = 'Bridge: invalid L1 origin';
+
+        const L1_ADDRESS_CANNOT_BE_ZERO: felt252 = 'Bridge: L1 address cannot be 0';
+        const L1_ADDRESS_OUT_OF_BOUNDS: felt252 = 'Bridge: L1 addr out of bounds';
+        const INVALID_RECIPIENT: felt252 = 'Bridge: invalid recipient';
+    }
+
+
+    // operation ID sent in the message payload to L1
+    const PROCESS_WITHDRAWAL: felt252 = 1;
+
+    // Ethereum addresses are bound to 2**160
+    const ETH_ADDRESS_BOUND: u256 = 0x10000000000000000000000000000000000000000_u256;
+
+    impl InitializableImpl = initializable_component::InitializableImpl<ContractState>;
+    impl InitializableInternalImpl = initializable_component::InternalImpl<ContractState>;
+
+    //
+    // Initializer
+    //
+
+    #[external(v0)]
+    #[generate_trait]
+    impl ERC20InitializerImpl of ERC20InitializerTrait {
+        fn initializer(ref self: ContractState, l1_bridge: felt252, l2_token: ContractAddress) {
+            assert(
+                self.world().is_owner(get_caller_address(), get_contract_address().into()),
+                Errors::CALLER_IS_NOT_OWNER
+            );
+
+            // one time bridge initialization
+            set!(
+                self.world(), DojoBridgeModel { token: get_contract_address(), l1_bridge, l2_token }
+            );
+
+            // reverts if already initialized
+            self.initializable.initialize();
+        }
+    }
+
+    //
+    // L1 Handler
+    //
+
+    #[l1_handler]
+    fn handle_deposit(
+        ref self: ContractState, from_address: felt252, recipient: ContractAddress, amount: u256
+    ) {
+        let data = self.get_data();
+        assert(from_address == data.l1_bridge, Errors::INVALID_L1_ORIGIN);
+
+        // mint token
+        let token_dispatcher = IERC20BridgeableDispatcher { contract_address: data.l2_token };
+        token_dispatcher.mint(recipient, amount);
+
+        let event = DepositHandled { recipient, amount };
+        self.emit_event(event);
+    }
+
+    //
+    // Impls
+    //
+
+    #[external(v0)]
+    #[generate_trait]
+    impl DojoBridgeImpl of DojoBridgeTrait {
+        fn initiate_withdrawal(ref self: ContractState, l1_recipient: felt252, amount: u256) {
+            let data = self.get_data();
+            let caller = get_caller_address();
+
+            assert(l1_recipient.is_non_zero(), Errors::L1_ADDRESS_CANNOT_BE_ZERO);
+            assert(l1_recipient.into() < ETH_ADDRESS_BOUND, Errors::L1_ADDRESS_OUT_OF_BOUNDS);
+            assert(l1_recipient != data.l1_bridge, Errors::INVALID_RECIPIENT);
+
+            // burn token
+            let token_dispatcher = IERC20BridgeableDispatcher { contract_address: data.l2_token };
+            token_dispatcher.burn(caller, amount);
```

Comment:
When messaging is involved, I would opt for escrow over burning directly. If anything goes wrong in the process, reverting an escrow is not affecting the liquidity the same way as burning/minting tokens.

But the escrow definitely adds more logic to handle, so having a process to mint them again is perhaps completely fine in this scenario.

Reply:
tokens are escrowed on L1 & mint/ burn on L2 & total_supply remains constant